### ISSUE TITLE
Deadlock and crash when quicky switching videos

### ISSUE THREAD

**Issue by HyperCubeUHD-4K** [ZeBobo5/Vlc.DotNet#544]
I have an issue about Vlc.DotNet.

# Generic information

- **Vlc.DotNet version** : 3.0.0
- **Vlc.DotNet project used** : WPF
- **libvlc version** : 3.0.6 x86/x64 get with nugget
- **.net version** : 4.5
- **Project language** : C#
- **Project build architecture** : AnyCPU
- **Operating system** : Windows 10 x64
- [x] I confirm that my issue doesn't happen in VLC itself.

Hello, I'm making a video player taking a list of video in a directory and open them when selected.
When I browse the list slowly, the player load the video, but when I start to go quickly, there is always a moment when the function play() start to deadlock my software.
I tried to put the function in a thread and instead of deadlocking, it shows me an error.

Here is a gif showing the problem

![image](https://i.imgur.com/Im0JDh5.gif)

More informations about the error seen in the gif :

![error](https://i.imgur.com/CS8NxbC.png)

and here the code that trigger the error

```c#
public string VideoPath
        {
            get
            {
                return _videoPath;
            }
            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                _videoPath = value;
                if(value != "" && File.Exists(value))
                {
                    progress_bar_timer.Stop();
                    
                    var thread = new System.Threading.Thread(delegate ()
                    {
                           //Deadlocking here when this is not in a thread and crashing when in one
                            VideoControl.SourceProvider.MediaPlayer.Play(new FileInfo(value));
                    });

                    thread.Start();
                }
                else
                {
                    var thread = new System.Threading.Thread(delegate () { VideoControl.SourceProvider.MediaPlayer.Stop(); });
                    thread.Start();
                }
                
            }
        }
```

Thanks !

**Comment by mfkl** [comment]
Does removing `[MethodImpl(MethodImplOptions.Synchronized)]` change anything?

**Comment by HyperCubeUHD-4K** [comment]
Thanks for your reply. 
I added this line because I thought it would fix the deadlock. 
I can't access the project right now, I will try to remove it this evening.

**Comment by HyperCubeUHD-4K** [comment]
The problem still occurs after removing the line :/

**Comment by dilandau2001** [comment]
I don't think it is a good practice to start threads in a property setter. I would avoid it.

Second, it seems you are using the same instance of player for every selection change, right?. How do you ensure in your code that when the thread 1 is calling start there is no other thread doing the stop?.

Maybe you should be creating and disposing a player each time or ensuring using locks that both things won't happen at the same time. 

Finally, my advise would be to also avoid loading the player that many times, don't start playing until the user maintain a selection for more than a second or so. You would avoid unnecessary cpu usage.

**Comment by HyperCubeUHD-4K** [comment]
I have followed your last advise and wait a little amount of time before the video to play with and it works perfectly, thanks.

I will try your other solutions for enhance my code.

**Comment by wutangyuan** [comment]
> I have followed your last advise and wait a little amount of time before the video to play with and it works perfectly, thanks.
> 
> I will try your other solutions for enhance my code.

How did you solve the problem? Can you share your solution？

**Comment by HyperCubeUHD-4K** [comment]
It's more a workaround than a proper solution but it works 🙂

First I initialize a timer and a counter variable

```C#

_counter_before_load_video = 2;

private void Init_Loading_Timer()
        {
            loading_timer = new System.Windows.Forms.Timer();
            loading_timer.Tick += new EventHandler(Check_Loading);
            loading_timer.Interval = 200; //in milliseconds
        }
```

This timer will call a function named `Check_Loading()` every 200 ms when it will be started

```C#
private void Check_Loading(object sender, EventArgs e)
        {
            Console.WriteLine("In timer");
            if (_counter_before_load_video <= 0)
            {
                Call_Video_Load();
                loading_timer.Stop();
            }
            else
            {
                _counter_before_load_video--;
            }
        }
```
Each time it enter this function, it will decrease the counter and when it reach 0, it will call the function `Call_Video_Load()` who will make a `raisePropertyChange("selectedVideo")` updating the player with the video selected

Now when we select an other video in the list, the attribute `Selected_Video` will be set with the new selected video in the list : 
```C#
public String Selected_Video
{
            get
            {
                return _selectedVideo;
            }
            set
            {
                _selectedVideo = value;
                Start_Loading_Timer();
            }
}
```

It call `Start_Loading_Timer()`, this function will reset the counter each time a new video is selected and start the timer if it's not started : 
```C#
private void Start_Loading_Timer()
        {
            _counter_before_load_video = 2;

            if (!loading_timer.Enabled)
            {
                loading_timer.Start();
            }
        }
```

In a nutshell, each time you select a new video in the list, these functions will make your program start the video after a 400ms idle in the last selected video. With this amount of time, the player can set itself properly without the crash I had and without being too annoying for the user.

I hope all of this will help you, I'm still here if you need more help and explanations 😄

**Comment by wutangyuan** [comment]
thank you very much for your reply.with your suggestions,I found that I
could solve the problems by modifying the code.

HyperCubeUHD-4K <notifications@github.com>于2020年5月21日 周四上午1:59写道：
